### ISSUE TITLE
Implement possibility to mark some nodes as not participating in leader election.

### ISSUE THREAD

**Issue by CyberDem0n** [zalando/patroni#12]

**Comment by jberkus** [comment]
This seems pretty simple, just another field of data for the replicas.

One question comes though about how those replicas push their recovery point to the SIS.  We also don't want other replicas to defer to them as the new master.  Thoughts?

**Comment by alexeyklyukin** [comment]
I’m looking into implementing this feature, in order to support another one, involving delayed replicas #36. IMO the non-voting member should continue returning its actual position. As an example, consider  a non-voting replica with a 1 hour delay: its lag from the master should be monitored the same as for the normal replica, only with a different delay limit to alert about. ATM the information about current xlog positions stored in DCS is not used by Patroni, which relies on the data obtained from the API call, but a monitoring system might decide to use the DCS data, as it can read the information about the whole cluster at once (at the expense of obtaining the information with up to loop_wait delay, which is 10 seconds by default).

In order for the other replicas to avoid deferring the master role to the non-voting one, we should expose the non-voting status in the information about the replica returned by the API, and use it during the check for the healthiest node, performed by each replica.

**Comment by alexeyklyukin** [comment]
Actually, this task is a subtask of #48.

**Comment by jberkus** [comment]
Alexey,

I thought we'd decided not to use the current xlog position in the DCS for anything?

Regardless, I think it's important that the extra tags for nodes be stored in the DCS, and that be the canonical source for "is this replica participating in master election or not".  The reason being that the  main use case is when you have 15+ replicas; we do NOT want to be polling 22 replicas to see which ones are participating; we want to poll ONLY the participating replicas.

This means that, on master election, replicas need to check their *own* DCS information in case of race conditions.

**Comment by jberkus** [comment]
Mind you, we should *also* expose it via the API.

**Comment by alexeyklyukin** [comment]
We are publishing the xlog positions for each replica in DCS, so the monitoring system might use it to evaluate the overall cluster health (i.e. are there nodes significantly behind the master). We are not using it in Patroni to make failover-related decisions, therefore, we won't have an issue with promotion of the 'wrong' replica into master based solely on the information from xlog position from DCS.

As for the tags, we definitely want to expose them via the API, because the ultimate decision to choose a certain node as a new master would be still performed after asking the API of that node. The question is whether we also want to *change* the tags via the API (without restarting Patroni)? I think the tags like clone- and replica- from might benefit from it, as it might be necessary to disable the replica as a source for base backups if there is a problem with it, or enable it only after making sure it is up-to-date with the master and running normally.

**Comment by jberkus** [comment]
What I'm saying is that those tags need to be published to the DCS, especially the nofailover tag.  Otherwise the feature does not help with large clusters.  The whole point of the feature is that, if I have a 22-server cluster, I don't want to (and can't wait for) poll the separate APIs of 21 other servers to find out which ones are failover-enabled.

Changing things via the API without restarting patroni needs to be its own issue discussion, I think.

**Comment by CyberDem0n** [comment]
Implemented in https://github.com/zalando/patroni/pull/80
